src/rebuild/initiator.c

```diff
@@ -353,6 +354,14 @@ rebuild_dkey(struct rebuild_tgt_pool_tracker *rpt,
 			rc = rebuild_fetch_update_bulk(rdone, oh, rebuild_cont);
 	}
 
+	if (rc == -DER_INPROGRESS) {
```

Comment:
Firstly, this section can be moved into above "if" branch.
Secondly, according to current rebuild logic, the in-rebuilding target only fetches data from the leader after the dtx_resync() done. If the leader return -DER_INPROGRESS, then related data should belong to some new DTX after the object rebuild start, such new modification will be or has been executed on the in-rebuilding target. So why need retry?

Reply:
I think I agree with you, probably in progress should not be checked here. I will think about it. Thanks.

Reply:
It turns out this EINPROGRESS comes from the oi record block,  so if it tries to insert the new oid into the oid table if the current oid table(record) is not committed yet, it will return -EINPROGRESS,  but this oid table allocation might comes from another transaction and different leader, so the local dtx resync will not be able to commit it before real rebuild.

Reply:
Is the -EINPROGRESS from ds_obj_fetch() or from local update?

Reply:
local update during vos_obj_hold()

Reply:
Such issue should be resolved with the commit 8f9661c7b0d19b21c1d4b780ec49834cab0ad9c3. So would you please to re-check the issue against the latest master?

Reply:
Not sure how this issue can be resolved by this patch, which seems only avoid the -EINPROGRESS for readonly operation? but this is write. Could you please explain a bit? thanks.

Reply:
Such patch removes the unnecessary availability check for modification case. That will cause fake DTX conflict as your case.

Reply:
Hmm, you mean this 			

if ((intent == DAOS_INTENT_UPDATE ||
     intent == DAOS_INTENT_PUNCH) &&
				    !(cmp & BTR_CMP_MATCHED))
					break;
It only skip the lookup without MATCHED flags?  But oi_find is using this flag
rc = dbtree_fetch(cont->vc_btr_hdl, BTR_PROBE_GE | BTR_PROBE_MATCHED,
                             intent, &key_iov, NULL, &val_iov);

Do I miss sth?

Reply:
You said you hit the issue when do the local update.

Reply:
yes, vos_update_end(0)->vos_obj_hold->vos_oi_find(), if I remember correctly.

Reply:
Then something for DTX availability check should be improved. I will double check and make patch if necessary.

Reply:
Please retry with the patch PR#880.

Reply:
I still do not understand why this can fix the issue.

The use case like this, suppose we have 4 servers, rank[0-3].

1. Rebuild starts, in the mean time, the normal I/O(DTX1) create a new oi entry on rank3
2. Then rebuild will also needs to insert new OID into the oi table, which will depends on the DTX1, so it might return EINPROGRESS.  I.e. rebuild only resync the uncommitted DTX before before,   and if the DTX created after rebuild starts, which is related with the ongoing rebuild, it might return EINPROGRESS.  

Could you please explain a bit more if I misunderstand sth?

So it has to retry for local update anyway.

Reply:
> I still do not understand why this can fix the issue.
> 
> The use case like this, suppose we have 4 servers, rank[0-3].
> 
> 1. Rebuild starts, in the mean time, the normal I/O(DTX1) create a new oi entry on rank3
> 2. Then rebuild will also needs to insert new OID into the oi table, which will depends on the DTX1, so it might return EINPROGRESS.  I.e. rebuild only resync the uncommitted DTX before before,   and if the DTX created after rebuild starts, which is related with the ongoing rebuild, it might return EINPROGRESS.

Not understand the situation, you mean that after the rebuild start, a new I/O request comes in, it will insert new OID in the object tree, right? Under such case, on the in-rebuilding target, both the I/O RPC handler and the rebuild ULT will try to insert the same new OID into the object tree, right?

Reply:
yes, if the I/O RPC handler is being handled first, then it will create a new entry in OID table tree, then the following rebuild ULT if it also tries to lookup the same OID( or insert the other OID in the same oid block), it will return EINPROGRESS?

Reply:
I think that the object rebuild logic only needs to handle the targets (object/key/record) with old epoch. For the ones generated after rebuild, why not skip them directly?

Reply:
This rebuild does only handle the object/key/record for old epoch, but it does not mean it will be executed before the normal I/O (new epoch I/O).  Unless dtx will commit by epoch?

Reply:
In fact, we have already recorded the timestamp when start the DTX resync:
ds_cont_child::sc_dtx_resync_time

Then any update with newer timestamp than such timestamp can be ignored by rebuild. If some DTX started before such timestamp, but not became committable before DTX resync start, then it will be either committed before DTX resync done, or aborted by force by the DTX resync. So when the logic moves to object rebuild, the in-processing IO will have newer epoch.

Reply:
hmm, you mean if rebuild has lower epoch than the on-going I/O, then EINPROGRESS will not return?  Sorry, not sure I follow.

Reply:
I mean that, for current implementation, during DTX resync, the DTX resync ULT will check the DTX epoch, if its epoch is newer than ds_cont_child::sc_dtx_resync_time, then such DTX is started after DTX resync triggered, so need NOT to be handed by DTX resync. The rebuild logic may also need similar check to skip new modification.

Then the left DTXs, their epochs are older than ds_cont_child::sc_dtx_resync_time, these DTXs will be either committed or aborted when DTX resync done. For the aborted ones, if related client/application is still alive, it will retry with newer epoch.